File: validator/auth/service-auth.cpp
#include "service-auth.h"
namespace tos::auth {
Result<bool> ServiceTrust::install_trusted(const ServicePolicy& policy, const ServiceKey& key) {
  if (policy.issuer_ == Hash{} || key.id == Hash{} || policy.suites_ != std::vector<Suite>{{1, 1}})
    return Error{"service-policy"};
  auto hash = object_id("service_policy", policy);
  if (!hash.ok())
    return hash.error();
  auto existing = history_.find(hash.value());
  if (existing != history_.end()) {
    if (existing->second.key.id != key.id || existing->second.key.public_key != key.public_key)
      return Error{"service-key-replacement"};
    return true;
  }
  auto previous = current_.find(policy.issuer_);
  if (previous == current_.end()) {
    if (policy.revision_ != 1 || policy.previous_ != Hash{})
      return Error{"service-policy-history"};
  } else {
    auto old = history_.find(previous->second);
    if (old == history_.end() || old->second.policy.revision_ == std::numeric_limits<std::uint64_t>::max() ||
        policy.revision_ != old->second.policy.revision_ + 1 || policy.previous_ != previous->second)
      return Error{"service-policy-history"};
  }
  auto admitted = AdmittedKey::admit(key.public_key);
  if (!admitted.ok())
    return admitted.error();
  history_.emplace(hash.value(), Installed{policy, hash.value(), key, admitted.value()});
  current_[policy.issuer_] = hash.value();
  return true;
}
Result<bool> ServiceTrust::verify_body(const Hash& issuer, const Hash& policy, std::span<const std::uint8_t> body,
                                       const std::vector<ServiceComponent>& components) const {
  auto found = history_.find(policy);
  if (found == history_.end() || found->second.policy.issuer_ != issuer)
    return Error{"service-issuer"};
  if (components.size() != 1)
    return Error{"service-components"};
  const auto& c = components[0];
  if (c.suite_ != 1 || c.parameters_ != 1 || c.key_id_ != found->second.key.id)
    return Error{"service-components"};
  auto signature = found->second.admitted.verify(body, c.signature_);
  if (!signature.ok())
    return signature.error();
  if (!signature.value())
    return Error{"service-signature"};
  return true;
}
Result<bool> ServiceTrust::verify(const Permit& permit) const {
  auto current = current_.find(permit.body_.issuer_);
  if (current == current_.end() || current->second != permit.body_.service_policy_)
    return Error{"stale-permit-policy"};
  auto raw = encode(permit.body_);
  if (!raw.ok())
    return raw.error();
  return verify_body(permit.body_.issuer_, permit.body_.service_policy_, raw.value(), permit.components_);
}
Result<bool> ServiceTrust::verify(const Receipt& receipt) const {
  auto raw = encode(receipt.body_);
  if (!raw.ok())
    return raw.error();
  return verify_body(receipt.body_.issuer_, receipt.body_.service_policy_, raw.value(), receipt.components_);
}
Result<bool> verify_permit(const Permit& permit, const PermitBody& expected, const ServiceTrust& trust,
                           std::uint32_t current, std::uint64_t fence, bool live) {
  if (permit.body_ != expected)
    return Error{"permit-context"};
  const auto& body = permit.body_;
  if (body.anchor_.seqno_ == std::numeric_limits<std::uint32_t>::max() || body.anchor_.root_ == Hash{} ||
      body.anchor_.file_ == Hash{} || body.anchor_.state_ == Hash{} ||
      body.expires_mc_ == std::numeric_limits<std::uint32_t>::max() || body.expires_mc_ < body.anchor_.seqno_ ||
      body.expires_mc_ - body.anchor_.seqno_ > 128)
    return Error{"permit-expiry"};
  if (current < body.anchor_.seqno_ || current > body.expires_mc_)
    return Error{"permit-current-coordinate"};
  if (fence == 0 || body.fence_ != fence)
    return Error{"fenced"};
  if (!live)
    return Error{"duty-not-permitted"};
  return trust.verify(permit);
}
Result<bool> verify_receipt(const Receipt& receipt, const ReceiptBody& expected, const ServiceTrust& trust,
                            const ReceiptWitness& witness) {
  const auto& body = receipt.body_;
  if (body != expected)
    return Error{"receipt-binding"};
  if (body.journal_sequence_ == 0 || body.fence_ == 0 || body.state_ < 1 || body.state_ > 3)
    return Error{"receipt-state"};
  auto checked = trust.verify(receipt);
  if (!checked.ok())
    return checked.error();
  auto id = object_id("receipt_body", body);
  if (!id.ok())
    return id.error();
  auto retained = witness.contains(body.journal_sequence_, id.value());
  if (!retained.ok())
    return retained.error();
  if (!retained.value())
    return Error{"receipt-frontier"};
  return true;
}
Result<bool> validate_request_state(const RequestState& state, const Hash& id) {
  auto raw = encode(state);
  if (!raw.ok())
    return raw.error();
  if (state.request_id_ != id)
    return Error{"state-correlation"};
  if (state.state_ > 3)
    return Error{"state-code"};
  if (state.state_ == 0) {
    if (state.statement_id_ != Hash{} || state.fence_ != 0 || !state.result_.empty() || !state.receipt_.empty())
      return Error{"absent-shape"};
    return true;
  }
  if (state.statement_id_ == Hash{} || state.fence_ == 0)
    return Error{"known-state"};
  if (state.state_ == 2) {
    if (state.result_.size() != 1 || !state.receipt_.empty())
      return Error{"complete-shape"};
    const auto& result = state.result_[0];
    if (result.request_id_ != id || result.statement_id_ != state.statement_id_ || result.fence_ != state.fence_)
      return Error{"state-result"};
  } else {
    if (!state.result_.empty() || state.receipt_.size() != 1)
      return Error{"pending-shape"};
    const auto& body = state.receipt_[0].body_;
    if (body.state_ != state.state_ || body.request_id_ != id || body.method_ != 5 ||
        body.subject_ != state.statement_id_ || body.fence_ != state.fence_ || body.result_hash_ != Hash{})
      return Error{"state-receipt-binding"};
  }
  return true;
}
}  // namespace tos::auth

File: tosctl/src/validator-auth/src/service_auth.rs
use crate::{
    codec::{encode, Error, Hash},
    crypto::{object_id, AdmittedKey},
    types::*,
};
use std::collections::BTreeMap;
#[derive(Clone, Debug)]
pub struct ServiceKey {
    pub id: Hash,
    pub public_key: Vec<u8>,
}
struct Installed {
    policy: ServicePolicy,
    key: ServiceKey,
    admitted: AdmittedKey,
}
#[derive(Default)]
pub struct ServiceTrust {
    history: BTreeMap<Hash, Installed>,
    current: BTreeMap<Hash, Hash>,
}
impl ServiceTrust {
    // Called only from independently authenticated local configuration.
    pub fn install_trusted(
        &mut self,
        policy: &ServicePolicy,
        key: &ServiceKey,
    ) -> Result<(), Error> {
        if policy.issuer == [0; 32]
            || key.id == [0; 32]
            || policy.suites != vec![Suite { suite: 1, parameters: 1 }]
        {
            return Err(Error("service-policy"));
        }
        let id = object_id("service_policy", policy)?;
        if let Some(old) = self.history.get(&id) {
            if old.key.id != key.id || old.key.public_key != key.public_key {
                return Err(Error("service-key-replacement"));
            }
            return Ok(());
        }
        if let Some(previous) = self.current.get(&policy.issuer) {
            let old = self.history.get(previous).ok_or(Error("service-policy-history"))?;
            if policy.previous != *previous
                || old.policy.revision.checked_add(1) != Some(policy.revision)
            {
                return Err(Error("service-policy-history"));
            }
        } else if policy.revision != 1 || policy.previous != [0; 32] {
            return Err(Error("service-policy-history"));
        }
        let admitted = AdmittedKey::admit(&key.public_key)?;
        self.history.insert(id, Installed { policy: policy.clone(), key: key.clone(), admitted });
        self.current.insert(policy.issuer, id);
        Ok(())
    }
    fn verify_body(
        &self,
        issuer: &Hash,
        policy: &Hash,
        raw: &[u8],
        components: &[ServiceComponent],
    ) -> Result<(), Error> {
        let found = self.history.get(policy).ok_or(Error("service-issuer"))?;
        if found.policy.issuer != *issuer {
            return Err(Error("service-issuer"));
        }
        if components.len() != 1 {
            return Err(Error("service-components"));
        }
        let c = &components[0];
        if c.suite != 1 || c.parameters != 1 || c.key_id != found.key.id {
            return Err(Error("service-components"));
        }
        if !found.admitted.verify(raw, &c.signature) {
            return Err(Error("service-signature"));
        }
        Ok(())
    }
    pub fn verify_permit(&self, value: &Permit) -> Result<(), Error> {
        if self.current.get(&value.body.issuer) != Some(&value.body.service_policy) {
            return Err(Error("stale-permit-policy"));
        }
        self.verify_body(
            &value.body.issuer,
            &value.body.service_policy,
            &encode(&value.body)?,
            &value.components,
        )
    }
    pub fn verify_receipt(&self, value: &Receipt) -> Result<(), Error> {
        self.verify_body(
            &value.body.issuer,
            &value.body.service_policy,
            &encode(&value.body)?,
            &value.components,
        )
    }
}
pub fn validate_permit_context(
    body: &PermitBody,
    current: u32,
    fence: u64,
    live: bool,
) -> Result<(), Error> {
    if [
        body.issuer,
        body.service_policy,
        body.audience,
        body.genesis_root,
        body.genesis_file,
        body.registry_root,
        body.policy,
        body.committee,
        body.session,
        body.identity,
        body.subject,
    ]
    .contains(&[0; 32])
        || ![4, 5, 7].contains(&body.method)
    {
        return Err(Error("permit-shape"));
    }
    if body.anchor.seqno == u32::MAX
        || [body.anchor.root, body.anchor.file, body.anchor.state].contains(&[0; 32])
        || body.expires_mc == u32::MAX
        || body.expires_mc.checked_sub(body.anchor.seqno).is_none_or(|v| v > 128)
    {
        return Err(Error("permit-expiry"));
    }
    if current < body.anchor.seqno || current > body.expires_mc {
        return Err(Error("permit-current-coordinate"));
    }
    if fence == 0 || body.fence != fence {
        return Err(Error("fenced"));
    }
    if !live {
        return Err(Error("duty-not-permitted"));
    }
    Ok(())
}
pub fn verify_permit(
    value: &Permit,
    expected: &PermitBody,
    trust: &ServiceTrust,
    current: u32,
    fence: u64,
    live: bool,
) -> Result<(), Error> {
    if value.body != *expected {
        return Err(Error("permit-context"));
    }
    validate_permit_context(&value.body, current, fence, live)?;
    trust.verify_permit(value)
}
pub trait ReceiptWitness {
    fn contains(&self, sequence: u64, receipt_body: &Hash) -> Result<bool, Error>;
}
pub fn verify_receipt(
    value: &Receipt,
    expected: &ReceiptBody,
    trust: &ServiceTrust,
    witness: &impl ReceiptWitness,
) -> Result<(), Error> {
    if value.body != *expected {
        return Err(Error("receipt-binding"));
    }
    let body = &value.body;
    if body.journal_sequence == 0 || body.fence == 0 || !(1..=3).contains(&body.state) {
        return Err(Error("receipt-state"));
    }
    trust.verify_receipt(value)?;
    if !witness.contains(body.journal_sequence, &object_id("receipt_body", body)?)? {
        return Err(Error("receipt-frontier"));
    }
    Ok(())
}
pub fn validate_request_state(value: &RequestState, id: &Hash) -> Result<(), Error> {
    encode(value)?;
    if value.request_id != *id {
        return Err(Error("state-correlation"));
    }
    if value.state > 3 {
        return Err(Error("state-code"));
    }
    if value.state == 0 {
        if value.statement_id != [0; 32]
            || value.fence != 0
            || !value.result.is_empty()
            || !value.receipt.is_empty()
        {
            return Err(Error("absent-shape"));
        }
    } else {
        if value.statement_id == [0; 32] || value.fence == 0 {
            return Err(Error("known-state"));
        }
        if value.state == 2 {
            if value.result.len() != 1 || !value.receipt.is_empty() {
                return Err(Error("complete-shape"));
            }
            let result = &value.result[0];
            if result.request_id != *id
                || result.statement_id != value.statement_id
                || result.fence != value.fence
            {
                return Err(Error("state-result"));
            }
        } else {
            if !value.result.is_empty() || value.receipt.len() != 1 {
                return Err(Error("pending-shape"));
            }
            let body = &value.receipt[0].body;
            if body.state != value.state
                || body.request_id != *id
                || body.method != 5
                || body.subject != value.statement_id
                || body.fence != value.fence
                || body.result_hash != [0; 32]
            {
                return Err(Error("state-receipt-binding"));
            }
        }
    }
    Ok(())
}
pub fn observe_request_state(
    previous: Option<&RequestState>,
    next: &RequestState,
    id: &Hash,
) -> Result<(), Error> {
    validate_request_state(next, id)?;
    if let Some(old) = previous {
        validate_request_state(old, id)?;
        if (old.state == 2 || old.state == 3) && old != next {
            return Err(Error("terminal-state-regression"));
        }
        if old.state == 1 {
            if next.state == 0 {
                return Err(Error("reserved-state-regression"));
            }
            if old.statement_id != next.statement_id || old.fence != next.fence {
                return Err(Error("reserved-state-binding"));
            }
        }
    }
    Ok(())
}
